File: src/timeline/c_api/workarea.cpp
/***

  Oak Video Editor - Non-Linear Video Editor
  Copyright (C) 2026 Oak Team

  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.

***/

#include "timeline/workarea.h"

#include <new>

#include "../src/timelineundoworkarea.h"
#include "../src/timelineworkarea.h"
#include "../../undo/c_api/commandhandle.h"

namespace
{

olive::TimelineWorkArea *impl(OakTimelineWorkArea *h)
{
	return reinterpret_cast<olive::TimelineWorkArea *>(h);
}

const olive::TimelineWorkArea *impl(const OakTimelineWorkArea *h)
{
	return reinterpret_cast<const olive::TimelineWorkArea *>(h);
}

OakUndoCommand *wrap_command(olive::UndoCommand *command)
{
	if (!command) {
		return NULL;
	}
	OakUndoCommand *handle = new (std::nothrow) OakUndoCommand{command, true};
	if (!handle) {
		delete command;
	}
	return handle;
}

olive::core::Rational rat(int64_t n, int64_t d)
{
	return olive::core::Rational(int(n), int(d));
}

} // namespace

OakTimelineWorkArea *oaktimeline_workarea_of(OakNodeNode *owner)
{
	if (!owner) {
		return NULL;
	}

	OakNodeWorkArea *workarea = NULL;
	if (oaknode_node_get_work_area(owner, &workarea) != OAKNODE_OK) {
		return NULL;
	}
	return reinterpret_cast<OakTimelineWorkArea *>(workarea);
}

int oaktimeline_workarea_get(const OakTimelineWorkArea *w, int *in_num,
							 int *in_den, int *out_num, int *out_den,
							 int *enabled)
{
	if (!w) {
		return OAKTIMELINE_E_INVALID;
	}

	if (in_num) {
		*in_num = impl(w)->in().numerator();
	}
	if (in_den) {
		*in_den = impl(w)->in().denominator();
	}
	if (out_num) {
		*out_num = impl(w)->out().numerator();
	}
	if (out_den) {
		*out_den = impl(w)->out().denominator();
	}
	if (enabled) {
		*enabled = impl(w)->enabled() ? 1 : 0;
	}
	return OAKTIMELINE_OK;
}

int oaktimeline_workarea_set_range(OakTimelineWorkArea *w, int in_num,
								   int in_den, int out_num, int out_den)
{
	if (!w) {
		return OAKTIMELINE_E_INVALID;
	}

	try {
		impl(w)->set_range(
			olive::core::TimeRange(rat(in_num, in_den), rat(out_num, out_den)));
		return OAKTIMELINE_OK;
	} catch (...) {
		return OAKTIMELINE_E_FAILED;
	}
}

OakUndoCommand *oaktimeline_workarea_set_range_command(
	OakTimelineWorkArea *w, int in_num, int in_den, int out_num,
	int out_den, int old_in_num, int old_in_den, int old_out_num,
	int old_out_den)
{
	if (!w) {
		return NULL;
	}

	try {
		olive::core::TimeRange range(rat(in_num, in_den), rat(out_num, out_den));
		olive::core::TimeRange old_range(rat(old_in_num, old_in_den),
										 rat(old_out_num, old_out_den));
		return wrap_command(
			new olive::WorkareaSetRangeCommand(impl(w), range, old_range));
	} catch (...) {
		return NULL;
	}
}

OakUndoCommand *oaktimeline_workarea_set_enabled_command(
	OakTimelineWorkArea *w, int enabled)
{
	if (!w) {
		return NULL;
	}

	try {
		return wrap_command(
			new olive::WorkareaSetEnabledCommand(impl(w), enabled != 0));
	} catch (...) {
		return NULL;
	}
}

int oaktimeline_workarea_reset(int *in_num, int *in_den, int *out_num,
							   int *out_den)
{
	if (!in_num || !in_den || !out_num || !out_den) {
		return OAKTIMELINE_E_INVALID;
	}

	*in_num = olive::TimelineWorkArea::k_reset_in.numerator();
	*in_den = olive::TimelineWorkArea::k_reset_in.denominator();
	*out_num = olive::TimelineWorkArea::k_reset_out.numerator();
	*out_den = olive::TimelineWorkArea::k_reset_out.denominator();
	return OAKTIMELINE_OK;
}

int oaktimeline_workarea_load(OakTimelineWorkArea *w, OakXmlReader reader)
{
	if (!w || !reader.ctx) {
		return OAKTIMELINE_E_INVALID;
	}

	olive::XmlStreamReader *native = oakcommon_xml_reader_get_native(reader);
	if (!native) {
		return OAKTIMELINE_E_INVALID;
	}

	try {
		return impl(w)->load(native) ? OAKTIMELINE_OK : OAKTIMELINE_E_FAILED;
	} catch (...) {
		return OAKTIMELINE_E_FAILED;
	}
}

int oaktimeline_workarea_save(const OakTimelineWorkArea *w,
							  OakXmlWriter writer)
{
	if (!w || !writer.ctx) {
		return OAKTIMELINE_E_INVALID;
	}

	olive::XmlStreamWriter *native = oakcommon_xml_writer_get_native(writer);
	if (!native) {
		return OAKTIMELINE_E_INVALID;
	}

	try {
		impl(w)->save(native);
		return OAKTIMELINE_OK;
	} catch (...) {
		return OAKTIMELINE_E_FAILED;
	}
}

File: crates/oaktimeline/src/workarea.rs
// Oak Video Editor - Non-Linear Video Editor
// Copyright (C) 2026 Oak Team
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

//! Timeline work area (`src/timeline/src/timelineworkarea.h` +
//! `timelineundoworkarea.h`): the in/out range, its enabled flag, and the
//! `WorkareaSetEnabledCommand`/`WorkareaSetRangeCommand` undo commands.
//!
//! De-Qt: no QObject, no signals — change notifications are the facade's
//! job.

use std::sync::{Arc, Mutex};

use oakcore_rs::{Rational, TimeRange};
use oakundo::undocommand::UndoCommand;

use crate::handle::get;

/// The reset sentinel `k_reset_in` (timelineworkarea.h): 0/1. Exposed as a
/// function because `Rational::new` is not yet `const` in oakcore-rs.
pub fn reset_in() -> Rational {
	Rational::new(0, 1)
}
/// The reset sentinel `k_reset_out` (timelineworkarea.h): RATIONAL_MAX,
/// i.e. 2147483647/1.
pub fn reset_out() -> Rational {
	Rational::new(2147483647, 1)
}

/// `TimelineWorkArea` — the in/out range on a timeline (timelineworkarea.h).
pub struct TimelineWorkArea {
	/// Whether the work area is enabled.
	workarea_enabled_: bool,
	/// The in/out range.
	workarea_range_: TimeRange,
}

impl TimelineWorkArea {
	/// A new, disabled work area at `k_reset_in`..`k_reset_out`.
	pub fn new() -> Self {
		Self {
			workarea_enabled_: false,
			workarea_range_: TimeRange::new(reset_in(), reset_out()),
		}
	}

	/// Whether the work area is enabled.
	pub fn enabled(&self) -> bool {
		self.workarea_enabled_
	}

	/// Set enabled.
	pub fn set_enabled(&mut self, e: bool) {
		self.workarea_enabled_ = e;
	}

	/// The in point.
	pub fn in_(&self) -> Rational {
		self.workarea_range_.in_()
	}

	/// The out point.
	pub fn out(&self) -> Rational {
		self.workarea_range_.out()
	}

	/// The range length.
	pub fn length(&self) -> Rational {
		self.workarea_range_.length()
	}

	/// The full range.
	pub fn range(&self) -> &TimeRange {
		&self.workarea_range_
	}

	/// Set the range.
	pub fn set_range(&mut self, range: TimeRange) {
		self.workarea_range_ = range;
	}
}

/// `WorkareaSetEnabledCommand` (timelineundoworkarea.h).
pub struct WorkareaSetEnabledCommand {
	/// Target work area (shared behind the facade's value handle; `None`
	/// for an empty/null handle, which makes the command a no-op).
	points: Option<Arc<Mutex<TimelineWorkArea>>>,
	/// New enabled flag.
	new_enabled: bool,
	/// Enabled flag captured at construction, restored by `undo`.
	old_enabled: bool,
}

impl WorkareaSetEnabledCommand {
	/// Construct from work area + new enabled value (captures old at ctor).
	pub fn new(points: crate::handle::CHandle, enabled: bool) -> Self {
		// SAFETY: work-area handles box an `Arc<Mutex<TimelineWorkArea>>`
		// (created by `make_owned`); reading it clones the shared `Arc`.
		let wa = unsafe { get::<Arc<Mutex<TimelineWorkArea>>>(&points) }.cloned();
		let old_enabled = wa
			.as_ref()
			.and_then(|w| {
				let w = w.lock().unwrap_or_else(|e| e.into_inner());
				Some(w.enabled())
			})
			.unwrap_or(false);
		Self {
			points: wa,
			new_enabled: enabled,
			old_enabled,
		}
	}

	/// Construct over an already-shared work area (the crate's Rust-typed
	/// entry; the facade-facing [`Self::new`] forwards here).
	pub fn new_arc(points: Arc<Mutex<TimelineWorkArea>>, enabled: bool) -> Self {
		let old_enabled = points.lock().unwrap_or_else(|e| e.into_inner()).enabled();
		Self {
			points: Some(points),
			new_enabled: enabled,
			old_enabled,
		}
	}

	/// `redo`: set enabled.
	pub fn redo(&mut self) {
		if let Some(wa) = &self.points {
			let mut wa = wa.lock().unwrap_or_else(|e| e.into_inner());
			wa.set_enabled(self.new_enabled);
		}
	}

	/// `undo`: restore old enabled.
	pub fn undo(&mut self) {
		if let Some(wa) = &self.points {
			let mut wa = wa.lock().unwrap_or_else(|e| e.into_inner());
			wa.set_enabled(self.old_enabled);
		}
	}

	/// Wrap as an oakundo command handle.
	pub fn to_command(self) -> UndoCommand {
		crate::undocommon::box_command(self)
	}
}

impl crate::undocommon::Command for WorkareaSetEnabledCommand {
	/// `Command::redo` — the inherent method takes precedence.
	fn redo(&mut self) {
		self.redo();
	}

	/// `Command::undo` — the inherent method takes precedence.
	fn undo(&mut self) {
		self.undo();
	}
}

/// `WorkareaSetRangeCommand` (timelineundoworkarea.h). The old range is
/// captured at construction when not supplied.
pub struct WorkareaSetRangeCommand {
	/// Target work area (shared behind the facade's value handle; `None`
	/// for an empty/null handle, which makes the command a no-op).
	workarea: Option<Arc<Mutex<TimelineWorkArea>>>,
	/// New range.
	new_range: TimeRange,
	/// Range captured at construction, restored by `undo`.
	old_range: TimeRange,
}

impl WorkareaSetRangeCommand {
	/// Construct from work area + new range (captures current as old).
	pub fn new(workarea: crate::handle::CHandle, range: TimeRange) -> Self {
		// SAFETY: as `WorkareaSetEnabledCommand::new`.
		let wa = unsafe { get::<Arc<Mutex<TimelineWorkArea>>>(&workarea) }.cloned();
		let old_range = wa
			.as_ref()
			.and_then(|w| {
				let w = w.lock().unwrap_or_else(|e| e.into_inner());
				Some(*w.range())
			})
			.unwrap_or(range);
		Self::new_with_old_opt(wa, range, old_range)
	}

	/// Construct over an already-shared work area (the crate's Rust-typed
	/// entry; the facade-facing [`Self::new`] forwards here).
	pub fn new_arc(workarea: Arc<Mutex<TimelineWorkArea>>, range: TimeRange) -> Self {
		let old_range = *workarea.lock().unwrap_or_else(|e| e.into_inner()).range();
		Self::new_with_old_opt(Some(workarea), range, old_range)
	}

	/// Construct from work area + new range + explicitly supplied old
	/// range (used by callers that may have a previously captured range;
	/// `new` delegates here with the current range).
	pub fn new_with_old(
		workarea: crate::handle::CHandle,
		range: TimeRange,
		old_range: TimeRange,
	) -> Self {
		// SAFETY: as `WorkareaSetEnabledCommand::new`.
		let wa = unsafe { get::<Arc<Mutex<TimelineWorkArea>>>(&workarea) }.cloned();
		Self::new_with_old_opt(wa, range, old_range)
	}

	/// Construct over an already-shared work area with an explicitly
	/// supplied old range (the crate's Rust-typed entry; the facade-facing
	/// [`Self::new_with_old`] forwards here).
	pub fn new_with_old_arc(
		workarea: Arc<Mutex<TimelineWorkArea>>,
		range: TimeRange,
		old_range: TimeRange,
	) -> Self {
		Self::new_with_old_opt(Some(workarea), range, old_range)
	}

	/// Shared constructor over an optional work area (`None` mirrors the
	/// empty/null handle: the command is a no-op).
	fn new_with_old_opt(
		workarea: Option<Arc<Mutex<TimelineWorkArea>>>,
		range: TimeRange,
		old_range: TimeRange,
	) -> Self {
		Self {
			workarea,
			new_range: range,
			old_range,
		}
	}

	/// `redo`: set the range.
	pub fn redo(&mut self) {
		if let Some(wa) = &self.workarea {
			let mut wa = wa.lock().unwrap_or_else(|e| e.into_inner());
			wa.set_range(self.new_range);
		}
	}

	/// `undo`: restore the old range.
	pub fn undo(&mut self) {
		if let Some(wa) = &self.workarea {
			let mut wa = wa.lock().unwrap_or_else(|e| e.into_inner());
			wa.set_range(self.old_range);
		}
	}

	/// Wrap as an oakundo command handle.
	pub fn to_command(self) -> UndoCommand {
		crate::undocommon::box_command(self)
	}
}

impl crate::undocommon::Command for WorkareaSetRangeCommand {
	/// `Command::redo` — the inherent method takes precedence.
	fn redo(&mut self) {
		self.redo();
	}

	/// `Command::undo` — the inherent method takes precedence.
	fn undo(&mut self) {
		self.undo();
	}
}
